hedera-node/hedera-app/src/test/java/com/hedera/node/app/blocks/impl/streaming/BlockNodeConnectionManagerTest.java

```diff
@@ -0,0 +1,127 @@
+    public void setUp() {
+        final var config = HederaTestConfigBuilder.create()
+                .withConfigDataType(BlockStreamConfig.class)
+                .withValue("blockStream.writerMode", "FILE_AND_GRPC")
+                .withValue("blockNode.blockNodeConnectionFileDir", "./src/test/resources/bootstrap")
+                .getOrCreateConfig();
+        given(mockConfigProvider.getConfiguration()).willReturn(new VersionedConfigImpl(config, 1));
+        blockNodeConnectionManager = new BlockNodeConnectionManager(mockConfigProvider);
+    }
+
+    @Test
+    void testNewBlockNodeConnectionManager() {
+        final var expectedGrpcEndpoint =
+                BlockStreamServiceGrpc.getPublishBlockStreamMethod().getBareMethodName();
+        assertEquals(expectedGrpcEndpoint, blockNodeConnectionManager.getGrpcEndPoint());
+    }
+
+    @Test
+    void testRetry_SuccessOnFirstAttempt() {
+        blockNodeConnectionManager.retry(mockSupplier, INITIAL_DELAY);
+
+        verify(mockSupplier, times(1)).get();
+        assertThat(logCaptor.infoLogs()).contains("Retrying in 10 ms");
+    }
+
+    @Test
+    void testRetry_SuccessOnRetry() {
+        when(mockSupplier.get())
+                .thenThrow(new RuntimeException("First attempt failed"))
+                .thenReturn(null);
+
+        blockNodeConnectionManager.retry(mockSupplier, INITIAL_DELAY);
+
+        verify(mockSupplier, times(2)).get();
+        assertThat(logCaptor.infoLogs()).contains("Retrying in 10 ms", "Retrying in 20 ms");
+    }
+
+    @Test
+    void testScheduleReconnect() throws InterruptedException {
+        blockNodeConnectionManager.scheduleReconnect(mockConnection);
+
+        verifyNoInteractions(mockConnection); // there should be no immediate attempt to establish a stream
+
+        Thread.sleep(BlockNodeConnectionManager.INITIAL_RETRY_DELAY.plusMillis(100));
+
+        assertThat(logCaptor.infoLogs()).contains("Retrying in 1000 ms");
+        verify(mockConnection, times(1)).establishStream();
+    }
+
+    @Test
+    void testEstablishConnections_PrioritizesNodes() {
+
+        // Establishing connections indirectly via waitForConnections
+        blockNodeConnectionManager.waitForConnections(Duration.ofSeconds(5));
+
+        List<String> infoLogs = logCaptor.infoLogs();
+        assertThat(infoLogs.get(0)).contains("Establishing connections to block nodes");
+
+        // Verify the order of connection attempts: The high priority node should be the first
+        assertThat(infoLogs.get(1)).contains("Connecting to block node localhost:8080");
+
+        // The lower priority nodes (any of node2, node3, node4) should come after that
+        assertThat(infoLogs.get(2))
+                .matches(log -> log.contains("Connecting to block node node2.example.com:8080")
+                        || log.contains("Connecting to block node node3.example.com:8081")
+                        || log.contains("Connecting to block node node4.example.com:8081"));
```

Comment:
Logging alone doesn't feel sufficient to verify this implementation logic is correct, but I'm having trouble thinking of something else at the moment 🤔